api/types/container/config.go

```diff
@@ -52,6 +53,16 @@ type HealthConfig struct {
 	Retries int `json:",omitempty"`
 }
 
+// SecretConfig holds self-sufficient data about a secret of a container.
+type SecretConfig struct {
+	ID     string      `json:",omitempty"` // Unique identifier of the secret
+	Data   []byte      `json:",omitempty"` // Raw content of the secret file
```

Comment:
Thanks for your consideration @thaJeztah.

> I don't think this is the right thing to do.

As I mentioned in the PR description, I totally agree that storing secret content as raw string in the container's configuration file is not the perfect so‍‍lution.

> Storing the secret this way would mean it doesn't provide benefits over using an `--env‍` ‍or mounting a file from the host inside the container.‍

But, I do not completely agree with this statement. I believe the concept of secrets has two main properties:

- They should be transmitted, and stored securely, as they might contain confidential data
- They should be configured to be mounted to the container with proper setup, i.e., destination file permissions and mode

This PR, although it has some security concerns around storing secrets, fixes the issue with the second point, which currently is not supported by Docker, as discussed in #40046.

Regarding the storage concern, as I realized, please correct me if I'm wrong, the current solution by Swarmkit is encrypting the snapshot of its raft state before being persisted, and decrypting the file during the startup process. To prevent adding the container's secrets into its `Config` property, I tried using `boltdb` in a similar solution to how the daemon stores information about volumes in `<daemon_root>/volumes/metadata.db` file, to `<daemon_root>/secrets.db` (see [this commit](https://github.com/xeptore/moby/commit/f65cafb2754f5819dc2324c3d9d9ae98e65befc7)), which simply stores a key-value pair of `containerID->secrets`. It gets loaded upon daemon startup, and closed/persisted at daemon shutdown phase. This solution has the benefit of not adding a new field to container `Config` field, and simplicity compared to moving toward `Swarmkit` in any way, but due to my lack of information about encryption/decryption, it actually stores secrets in a raw format. I'm also open to any further research or implementing a solution around dealing with secret storage, if you'd guide me.

Reply:
Swarm secrets are stored in swarm's raft store (which is encrypted), and (but it's been a while since I looked at the full implementation), it uses a `memfs` mount to prevent the mounted file from being written to disk.
For transmitting the data, we should also make sure that it doesn't end up in logs (for example, with the daemon in debug mode, API request calls may be logged).

That said; I don't think using the SwarmKit store for this is ideal; using the SwarmKit store means that secrets will always have to be "imported" into that store (which somewhat makes sense for swarm services, to distribute them, but for situations where a secure store is already available for secrets, means that the secret is duplicated); I know some work was done to allow swarm services to use an external store (see https://github.com/moby/swarmkit/pull/2326), so possibly something like that should be explored as an option (and make it available for both "swarm" and "non-swarm" containers, allowing `docker secret` and `docker config` to be used for both).

Reply:
What do you think about using an encryption algorithm similar to what Swarmkit is doing to encrypt its raft snapshots, then using a `memfs` or `ramfs` mount typed storage with the same simple storage mechanism, e.g., `boltdb`, implementation for daemon's local container secrets, instead of making changes to Swarmkit?